server/jetty/src/main/java/io/deephaven/server/jetty/JsPluginsZipFilesystem.java

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright (c) 2016-2023 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.server.jetty;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.deephaven.configuration.CacheDir;
+import io.deephaven.plugin.js.JsPluginManifestPath;
+import io.deephaven.plugin.js.JsPluginPackagePath;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.URI;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+class JsPluginsZipFilesystem {
+    private static final String ZIP_ROOT = "/";
+
+    /**
+     * Creates a new js plugins instance with a temporary zip filesystem.
+     *
+     * @return the js plugins
+     * @throws IOException if an I/O exception occurs
+     */
+    public static JsPluginsZipFilesystem create() throws IOException {
+        final Path tempDir =
+                Files.createTempDirectory(CacheDir.get(), "." + JsPluginsZipFilesystem.class.getSimpleName());
+        tempDir.toFile().deleteOnExit();
+        final Path fsZip = tempDir.resolve("deephaven-js-plugins.zip");
+        fsZip.toFile().deleteOnExit();
+        final URI uri = URI.create(String.format("jar:file:%s!/", fsZip));
+        final JsPluginsZipFilesystem jsPlugins = new JsPluginsZipFilesystem(uri);
+        jsPlugins.init();
+        return jsPlugins;
+    }
+
+    private final URI filesystem;
+    private final List<JsPluginManifestEntry> entries;
+
+    private JsPluginsZipFilesystem(URI filesystem) {
+        this.filesystem = Objects.requireNonNull(filesystem);
+        this.entries = new ArrayList<>();
+    }
+
+    public URI filesystem() {
+        return filesystem;
+    }
+
+    public synchronized void copyFrom(JsPluginPackagePath srcPackagePath, JsPluginManifestEntry srcEntry)
+            throws IOException {
+        checkExisting(srcEntry);
+        final Path srcDist = srcPackagePath.distributionPath(srcEntry.main());
+        // TODO(deephaven-core#3005): js-plugins checksum-based caching
+        try (final FileSystem fs = FileSystems.newFileSystem(filesystem, Map.of())) {
+            final JsPluginManifestPath manifest = manifest(fs);
+            final Path destDist = manifest
+                    .packagePath(srcEntry.name())
+                    .distributionPath(srcEntry.main());
+            CopyHelper.copyRecursive(srcDist, destDist);
+            entries.add(srcEntry);
+            writeManifest(fs);
```

Comment:
If I understand this correctly, each copyFrom call re-opens the zip file, appends new files to it, then appends a new ~manifest~ central directory entry to it, and writes a new toc at the end of the file, and closes it.

The reason for doing this is that we then pass the path to this zip directly to jetty, so it can open it to use as its resource base, and serve files from it. Jetty prefers to not serve from zips, but we disable that to avoid unpacking to /tmp and risk something cleaning out that directory. So we're finding already-unpacked contents on disk so we can pack it and tell jetty to server it still-packed... 

Tradeoffs: 
 * the contents are immutable (changes won't break the server), but the contents are immutable (harder to develop live)
 * extra copy on disk (deleted on exit, assuming non-crash exit)
 * contents served from zip (shouldn't be a big deal, assuming good caching, small files)
 * contents ostensibly filtered to only what is expected to be served

We don't support reading zips/tarballs (i.e. jars, wheels, npm packages), and so always already have an unpacked structure on disk somewhere, and we require (i _think_) that the unpacked structure matches what we will zip (though we only copy/serve a subset of what was in that directory). Why not point to those existing unpacked dirs, with some filtering for "you only need to see these files"?

Reply:
> Jetty prefers to not serve from zips, but we disable that

Is this documented somewhere? I don't think we need to disable anything to make jetty work with zips...

> Why not point to those existing unpacked dirs, with some filtering for "you only need to see these files"?

This _might_ be possible, but it's somewhat dependent on being able to configure jetty later in our startup lifecycle (or, get the plugin information earlier in our lifecycle) than we currently are.

Right now, the zip represents the single-point of entry that we can later add to via plugins:

```java
        // Wire up /js-plugins/*
        context.addServlet(zipFsServlet, "/js-plugins/*");
```

If we were to allow plugins to add to them, it could look something like:

```java
        WebAppContext context = new WebAppContext(null, "/js-plugins/<pluginName>", null, null, null, new ErrorPageErrorHandler(), NO_SESSIONS);
        context.setBaseResource(createResource(<pluginPath>));
        addHandler.accept(context);
```

And then we'd need to create an immutable manifest.json resource as well.

I think it would be a tricky change to the order of when things are created during startup, ala `io.deephaven.server.runner.DeephavenApiServer#run`.

Reply:
Of course, there may be ways we could use a `Servlet` and add later w/ with uncopied plugin Resources, but I don't really know the specific jetty machinations to see if that would be possible.